[optim] FusedAdam/W accepts lr: Tensor without h2ds

janeyx99 opened the pull request:
Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106916

janeyx99 committed:
[wip] init adam fused lr

[ghstack-poisoned]

janeyx99 committed:
Update on "[wip] init adam fused lr"




[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Deletes unneeded .pyi files for adam and adamw
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I could tease out those changes, but eh

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Deletes unneeded .pyi files for adam and adamw
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I could tease out those changes, but eh

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)





[ghstack-poisoned]

janeyx99 commented:
The broken tests are a result of the torch.distributed.optim._functional* optimizers diverging from our new broader schema and torch.jit.script really disliking the schema change.

I'm curious whether the distributed functional optims are still used and if they still need to be jit scriptable. cc @wanchaol

albanD commented on torch/optim/adam.pyi:
```diff
@@ -1,20 +1,22 @@
-from typing import Optional, Tuple
```
You added types inline above. So this file can be deleted right?

albanD commented on torch/optim/adamw.pyi:
```diff
@@ -1,20 +1,22 @@
-from typing import Optional, Tuple
```
Same to remove

albanD reviewed (commented):
I can't find the right place in the code. But you're missing checks that the lr is on the right device no?

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)





[ghstack-poisoned]

janeyx99 commented:
> I can't find the right place in the code. But you're missing checks that the lr is on the right device no?

In _fused_adam I move them

albanD commented on test/test_proxy_tensor.py:
```diff
@@ -814,7 +814,7 @@ def fused_adam(params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs, state_step
                 max_exp_avg_sqs,
                 state_steps,
-                lr=0.1,
+                lr=torch.tensor(0.1),
```
Both should work no?

janeyx99 replied:
This calls the private API directly so it does not pass the schema check. i.e., it fails to compile and would not work.

albanD commented on torch/optim/adam.py:
```diff
@@ -108,6 +118,14 @@ def _init_group(
+                if group['fused'] and not torch.is_tensor(group['lr']):
+                    group['lr'] = torch.tensor(group['lr'], dtype=torch.float, device=p.device)
+                elif not group['fused'] and torch.is_tensor(group['lr']):
+                    group['lr'] = group['lr'].item()
```
We do this because our fop-loop and foreach code actually raises an error when given a tensor()? Do we test this?

janeyx99 replied:
We do this because it's better to set the lr correctly globally when one calls step through the optimizer. The for-loop and foreach code should never raise user-level errors though because we coerce the types in init_group and adam() which are the exposed endpoints.

But yes, agreed, I should add tests for passing tensors as lr for the forloop and foreach impls too.

albanD replied:
So today this `.item()` call happens anyways (explicitly or implicitly) during the step() right?

janeyx99 replied:
only if the user calls adam directly  as a functional api. if the user inits optimizers through the constructor, the .item is only called during the first step in init_group.

we could have our forloop and foreach case error instead if we want to be blatant about the behavior

janeyx99 replied:
The .item only happens in PT2 so tehre's no need to blatantly coerce in eager

albanD commented on torch/optim/adam.py:
```diff
@@ -278,10 +296,19 @@ def adam(params: List[Tensor],
     if fused and not torch.jit.is_scripting():
+        # lr needs to be a tensor
+        if not isinstance(lr, Tensor):
+            lr = torch.tensor(lr, dtype=torch.float)
```
Why aren't these errors? BC of checkpoints?

janeyx99 replied:
Sure, but also because people are allowed to call adam() directly.

albanD replied:
But this might end up on the wrong device?
In general, I would say that the minimum cross device moves code we have, the better. Unless it would be breaking current behavior, I would not enable this.

janeyx99 replied:
If we don't do this, we will definitely be breaking the current behaviour for anyone calling

`adam.adam(params, grads, ..., lr: float=whatever, fused=True)`

albanD commented on torch/optim/adam.py:
```diff
@@ -315,14 +342,15 @@ def _single_tensor_adam(params: List[Tensor],
                         differentiable: bool):
 
     assert grad_scale is None and found_inf is None
+    assert not isinstance(lr, Tensor), "lr should have been coerced to a scalar in _single_tensor_adam"
```
nit: in -> before calling

janeyx99 replied:
Sorry I don't know how to interpret your recommendation 😬

albanD commented on torch/optim/adam.py:
```diff
@@ -428,6 +456,8 @@ def _multi_tensor_adam(params: List[Tensor],
     if len(params) == 0:
         return
 
+    assert not isinstance(lr, Tensor), "lr should have been coerced to a scalar in _multi_tensor_adam"
```
before calling

albanD commented on torch/optim/adam.py:
```diff
@@ -557,19 +587,22 @@ def _fused_adam(
+        if not torch.is_tensor(lr):
+            lr = torch.tensor(lr, device=device, dtype=torch.float)
+        elif lr.device != device:
+            lr = lr.to(device=device)
```
Cross device?? Most likely no no?

janeyx99 replied:
Why is this bad? Don't we do something similar for grad_scale and found_inf?

janeyx99 replied:
Revised impl to not need this by adding an overload

albanD commented on torch/optim/adam.py:
```diff
@@ -557,19 +587,22 @@ def _fused_adam(
+                       device_max_exp_avg_sqs,
+                       device_state_steps,), _) in grouped_tensors.items():
+        if not torch.is_tensor(lr):
+            lr = torch.tensor(lr, device=device, dtype=torch.float)
```
Why not error?

janeyx99 replied:
We can error

albanD reviewed (commented):
haven't looked at adamw

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

albanD commented on torch/optim/adam.py:
```diff
@@ -557,19 +589,21 @@ def _fused_adam(
     if not params:
         return
+    if differentiable:
+        raise RuntimeError("_fused_adam is not differentiable")
```
Should this be `Adam` with `fused=True` does not support `differentiable=True` ?

janeyx99 replied:
ok

albanD commented on torch/optim/adam.py:
```diff
@@ -557,19 +589,21 @@ def _fused_adam(
+
     grad_scale_dict = {grad_scale.device: grad_scale} if grad_scale is not None else None
     found_inf_dict = {found_inf.device: found_inf} if found_inf is not None else None
+    lr_dict = {lr.device: lr} if isinstance(lr, Tensor) and lr.device != "cpu" else None
```
This deserves a comment I think!

janeyx99 replied:
added

albanD reviewed (commented):
Sounds good!

janeyx99 commented on aten/src/ATen/native/cuda/FusedAdamKernel.cu:
```diff
@@ -42,4 +42,35 @@ void _fused_adam_kernel_cuda_(
+) {
+  if (amsgrad) {
+    TORCH_CHECK(
+        at::native::check_fast_path_restrictions({params, grads, exp_avgs, exp_avg_sqs, max_exp_avg_sqs}),
```
for foreachhhhhh.

I'm not going to address this naming in this PR as I believe it is out of scope

albanD reviewed (approved):
All my comments are nits so good to go once the phrasing makes everyone happy!

mlazos commented on torch/optim/adam.py:
```diff
@@ -108,6 +120,11 @@ def _init_group(
                 if group['differentiable'] and state['step'].requires_grad:
                     raise RuntimeError('`requires_grad` is not supported for `step` in differentiable mode')
+
+                # Foreach without capturable does not support a tensor lr
```
why does capturable impact whether or not a tensor LR is supported in foreach? wouldn't it work just not be performant?

janeyx99 replied:
The capturable path allows lr to be a tensor cuz we have the overloads for those ops. The non capturable path uses python math instead as that is more performant without needing as much device memory. An easy way to "support" tensor LR is just to autoswitch to capturable when we discover that lr is a tensor, but that's not necessarily a decision i want to make in this PR.

mlazos commented on torch/optim/adam.py:
```diff
@@ -1,22 +1,34 @@
-                 differentiable: bool = False, fused: Optional[bool] = None):
+    def __init__(self,
+                 params: params_t,
+                 lr: Union[float, Tensor] = 1e-3,
```
Does passing a single tensor here support multi device? shouldn't it be treated like the other scalars and be placed into an array?

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on " [optim] FusedAdam/W accepts lr: Tensor without h2ds"

Starts addressing #106802

This PR also conveniently does some BE:
- Fixes a bug in adamw where we use amsgrad instead of per group amsgrad
- Brings the impls of adamw and adam closer to correctness and to each other

I couldn't fully remove the .pyi's because mypy was going to complain about the entire files which scared me and shouldn't go in this PR anyway.

Test plan:
- Add tests to ensure that lr could be passed as a Tensor 
- Did some profiling of the below code (runs 1k iterations of step for Adam)

```
import torch
from torch.testing._internal.common_utils import TestCase

param = torch.rand(2, 3, dtype=torch.float, device='cuda:0', requires_grad=True)
param.grad = torch.rand_like(param)

lr = torch.tensor(.001, device='cuda:0')
opt = torch.optim.Adam([param], lr=lr, fused=True)

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ]
) as p:
    for _ in range(1000):
        opt.step()

print(p.key_averages().table(sort_by="cpu_time_total"))

```

Before my change:
<img width="1381" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/cfc5175a-0f41-4829-941f-342554f3b152">


After my change (notice there are no d2h syncs and the CPU time is lower!):
![image](https://github.com/pytorch/pytorch/assets/31798555/726d7e66-dcff-4a4f-8a75-e84329961989)

Next steps long term:
- have all capturable foreach + forloop impls in Adam(W) handle tensor LR
- have all capturable impls handle tensor LR
- have all impls handle tensor LR





[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge
